src/resource/resource_package.rs

```diff
@@ -28,30 +28,32 @@ pub enum ResourcePackageError {
 }
 
 #[allow(dead_code)]
-#[derive(BinRead)]
-#[br(import(is_patch: bool))]
+#[binrw]
+#[brw(little, import(is_patch: bool))]
 pub struct ResourcePackage {
-    magic: [u8; 4],
+    pub magic: [u8; 4],
 
-    #[br(if (magic == * b"2KPR"))]
-    metadata: Option<PackageMetadata>,
+    #[br(if (magic == *b"2KPR"))]
+    #[bw(if (magic == b"2KPR"))]
+    pub metadata: Option<PackageMetadata>,
 
-    header: PackageHeader,
+    pub header: PackageHeader,
 
-    #[br(if (is_patch && header.table_offset != 0 && metadata.as_ref().map_or(true, | m | m.patch_id > 0)))]
-    unneeded_resource_count: u32,
+    #[brw(if (is_patch))]
+    pub unneeded_resource_count: u32,
 
-    #[br(if(is_patch))]
-    #[br(little, count = unneeded_resource_count, map = |ids: Vec<u64>| {
+    #[brw(if (is_patch))]
+    #[br(count = unneeded_resource_count, map = |ids: Vec<u64>| {
     match unneeded_resource_count{
         0 => None,
         _ => Some(ids.into_iter().map(RuntimeResourceID::from).collect::<Vec<_>>()),
     }
     })]
-    unneeded_resources: Option<Vec<RuntimeResourceID>>,
+    pub unneeded_resources: Option<Vec<RuntimeResourceID>>,
 
     #[br(parse_with = resource_parser, args(header.file_count))]
-    resources: HashMap<RuntimeResourceID, ResourceInfo>,
+    #[bw(write_with = empty_writer)]
+    pub resources: HashMap<RuntimeResourceID, ResourceInfo>,
```

Comment:
I haven't cleaned up the access modifiers yet. Will do so in a follow up commit.

Reply:
These have now been adjusted. Fields that don't need to be public have been made crate public. Fields that were being returned as-is from member methods have been made / kept public and relevant methods have been removed, since there is no need for an extra layer of indirection.

Reply:
I've also made it so the ResourcePackage keeps track of its source internally (eg. the file path it was read from) so the `read_resource` no longer needs a path passed to it explicitly and there's no way to misuse it (by passing the wrong path for example).